### PR TITLE
Enable immutability in Key class.

### PR DESCRIPTION
**Pull request by vpatryshev**
And added KeyTest.

This is a small change.

---
This change is [Reviewable](https://reviewable.io/reviews/h2oai/h2o-3/302)

### CODE REVIEW

File: h2o-core/src/main/java/water/Key.java [hunk 1]
```diff
@@ -467,12 +472,13 @@ public String valueClassSimple() {
...
+    return k != null && _hash == k._hash && Arrays.equals(k._kb,_kb);
+  }
+
   @Override public boolean equals( Object o ) {
-    if( this == o ) return true;
-    if( o == null ) return false;
-    Key k = (Key)o;
-    if( _hash != k._hash ) return false;
-    return Arrays.equals(k._kb,_kb);
+    return this == o || (o != null && (o instanceof Key) && equals((Key)o));
```
Comment (st-pasha) [inline h2o-core/src/main/java/water/Key.java]: I don't quite understand this. If `o` is a `Key` then the compiler should have called the `public boolean equals(Key k)` method above. Is there a case when this method returns true then?

Comment (vpatryshev) [inline h2o-core/src/main/java/water/Key.java]: Yes, of course, it's supposed to call the other equals. I just don't count on it; no big deal.
This method will return true if called like this:
    Object key3 = Key.make(s1);
    assertTrue(key1.equals(key3));

**Comment by cliffclick** [comment]
fyi - this change looks crazy bad.

Key compares happen a zillion times a second.  Any bulk data operation 
not specifically bulk-optimized will flood the system with Keys... all 
needing key-compares.

The hash equals check needs to happen before the array equals.


Cliff

**Comment by mmalohlava** [comment]
@vpatryshev is there any reported bug/misbehavior? Right now without performance tests, even minor change inside the core infrastructure can have significant impact on performance...  
thx Cliff to keeping eye on it.

**Comment by vpatryshev** [comment]
Cliff, did you see the code? Have you seen this:
 public boolean equals(Key k) {
    return k != null && _hash == k._hash && Arrays.equals(k._kb,_kb);
 }

That's exactly what you were looking for, right? No reason to compare arrays if hashes are different.

**Comment by vpatryshev** [comment]
Michal, performance does not suffer here. It's the same sequence of operations. The code is more compact and less exotic, more typical compared to a bunch of other Java classes.

**Comment by vpatryshev** [comment]
Guys, we probably can get together tomorrow and discuss all the issues with this code, ok?

**Comment by cliffclick** [comment]
I have free time today.  Cell 408-307-0321

Cliff

**Comment by cliffclick** [comment]
Code reviewing now.

I see a bunch of issues in 30sec of looking - clear perf killers (clone 
instead of returning the same array, in hot code), obfuscation, nested 
non-final function calls where none existed before, function calls 
instead plain constant array lookups, etc, etc, etc.


Cliff

**Comment by vpatryshev** [comment]
Will work on making sure cloning is not blindly used on all occasions. Thanks.

**Comment by st-pasha** [comment]
Agree with @mmalohlava that it is very desirable to have performance tests for such critical data structures; however we can't assume that the original code was written in the optimal way either (by the same reason: there were no performance tests).

However I don't see anything "crazy bad" here either. There are few minor performance improvements: in lines 71, 75, 85, 180-183. Line 276 is minor performance hit for the benefit of increased safety (if the client passes a byte array and then changes that byte array afterwards, we'll get into trouble). Line 439 (replacing an array lookup with a function call) has no performance implication because JIT will inline it anyways. Line 476 (the one that is to be used "zillion times") is actually performance boost: it skips one case of Object to Key for those cases when the compiler can statically determine that (which is like 99.9% of cases). The second version of the function `equals(Object)` will be used extremely rarely, and hence its performance is irrelevant. Finally, the `final` modifier on a static method is meaningless and can be safely removed.

There are still ways in which this PR can be improved of course, but those should be better addressed on a line-by-line basis.

File: h2o-core/src/main/java/water/Key.java [hunk 2]
```diff
@@ -255,14 +252,29 @@ private Key(byte[] kb) {
...
+  static <T extends Keyed> Key<T> buildKeyForTestingPurposes(byte[] kb, int hash) {
+    return new Key<>(kb, hash);
+  }
+
   // Make new Keys.  Optimistically attempt interning, but no guarantee.
   static <P extends Keyed> Key<P> make(byte[] kb, byte rf) {
     if( rf == -1 ) throw new IllegalArgumentException();
-    Key key = new Key(kb);
-    Key key2 = H2O.getk(key); // Get the interned version, if any
+    Key<P> key = new Key<>(kb.clone());
```
Comment (st-pasha) [inline h2o-core/src/main/java/water/Key.java]: For an arbitrary `byte[]` array making a clone is the right thing to do. However often we make a Key from a String (which is then cast to byte[]) -- there should be a version of the constructor that optimizes that use case.

File: h2o-core/src/main/java/water/Key.java [hunk 3]
```diff
@@ -467,12 +472,13 @@ public String valueClassSimple() {
   }
 
   @Override public int hashCode() { return _hash; }
+
+  public boolean equals(Key k) {
```
Comment (st-pasha) [inline h2o-core/src/main/java/water/Key.java]: could be made final for tiny performance boost

**Comment by mmalohlava** [comment]
Can you please discuss that with @tomasnykodym (and with Cliff if he is available, i am on the way to cust)? Any minor performance overhead can have huge impact at this level of API.

**Comment by cliffclick** [comment]
On 10/6/2016 11:00 AM, Pasha Stetsenko wrote:
>
> Agree with @mmalohlava <https://github.com/mmalohlava> that it is very 
> desirable to have performance tests for such critical data structures; 
> however we can't assume that the original code was written in the 
> optimal way either (by the same reason: there were no performance tests).
>
Yes, there was lots and lots of performance tests - and the code was 
very specifically written for speed in the critical structures. *And 
your assumption that they were not is itself suspicious: *H2O has a rep 
for performance exactly because it was built for speed from the git-go, 
and exactly because I pushed hard for 4 yrs to keep it fast.

All changes here should be viewed with heavy suspicion vis-a-vis 
performance.

In particular, either version of this code will run "fast enough"... 
until it doesn't.  Like building codes, this code is written in blood.  
Over the years all these lines of code have been optimized *because* 
they become *the hot bottleneck* in some crucial case or another.

I'll say it again: changes here should be viewed with heavy suspicion.


> However I don't see anything "crazy bad" here either. There are few 
> minor performance improvements: in lines 71, 75, 85, 180-183. Line 276 
> is minor performance hit for the benefit of increased safety (if the 
> client passes a byte array and then changes that byte array 
> afterwards, we'll get into trouble).
>
Never happened in 4 yrs.  You're guarding against a non-event, *at the 
cost of an understandable cost-model*.  Been there, done that for years, 
and it's not worth it.  As people lose contact with *why* the code is 
fast, they make more and more performance mistakes.  A year or two goes 
by, everthing's 50% slower and know one knows why.


> Line 439 (replacing an array lookup with a function call) has no 
> performance implication because JIT will inline it anyways.
>
Nope - it *mostly* will inline.  Until it doesn't.  Been there, wrote 
the inliner, been hauled up in front of senior Sun execs because some 
hot loop in a customer code triggered the too-deep inliner cutoff and 
went 10x slower.  Lesson is the same as above: in hot code, don't assume 
perfect JIT behavior.  Manually inline.


> Line 476 (the one that is to be used "zillion times") is actually 
> performance boost: it skips one case of Object to Key for those cases 
> when the compiler can statically determine that (which is like 99.9% 
> of cases). The second version of the function |equals(Object)| will be 
> used extremely rarely, and hence its performance is irrelevant.
>
Actually.. the equals(Object) is probably by far and away the most 
common usage - because Keys go into a generic hashmap (NBHM usually, but 
same principle applies in any case) for which the strong type is NOT 
known at the bytecode level.  Generics don't help; the JVM doesn't see 
those types (and can't trust them if it could, hence it must do the 
type-check anyways).


> Finally, the |final| modifier on a static method is meaningless and 
> can be safely removed.
>
Agreed.

**Comment by cliffclick** [comment]
Again, I am available today.  Call me, or skype or slack or whatever.
408-307-0321

Cliff

File: h2o-core/src/main/java/water/Key.java [hunk 4]
```diff
@@ -416,22 +429,14 @@ public String valueClassSimple() {
...
+      if (!((' ' <= a && a <= '#') || ('%' <= a && a <= '~'))) break;
     }
     if (len>=0) {
       StringBuilder sb = new StringBuilder();
       sb.append(MAGIC_CHAR);
       for( int i = 0; i <= len; ++i ) {
         byte a = _kb[i];
-        sb.append(HEX[(a >> 4) & 0x0F]);
-        sb.append(HEX[(a >> 0) & 0x0F]);
+        sb.append(hex(a >> 4)).append(hex(a));
```
Comment (st-pasha) [inline h2o-core/src/main/java/water/Key.java]: Perhaps this line can be further sped up by replacing it with `sb.append(HEXNEW, a*2, 2)`, where  `static char[] HEXNEW = "00010203......fdfeff".toCharArray()`

**Comment by cliffclick** [comment]
:-)

toString() shouldn't be on any hot path.  Here I'll take clarity over 
raw speed hacks.

2-at-a-shot nybble conversions might be faster (might not: your lookup 
array got a LOT bigger) but I don't think they are any clearer.  Guess 
I've done a lot of byte/nybble math hacks over the years.

However, "hex()" vs "HEX[]" - the plain array lookup is more clear to 
me, especially since it's from a static constant array.

In C there's a more clear expression that doesn't quite pass the Java 
syntax gods:
|  *str++ = "0123456789ABCDEF"[(a>>4)&0x0F];|

I had to "outline" the constant String and move it out of the function, 
and I couldn't actually use a String-syntax for the desired byte-array 
of ASCII chars.

FYI, unlike the inlining rules, the array math will 100% optimize down 
to the minimal bits.  So crud like byte[(a>>4)&0x0F] emits machine code 
exactly how you might expect it does.


Cliff

**Comment by vpatryshev** [comment]
The justification here is making sure immutability makes the code safer and
does not lose a drop of performance. We can drop this specific PR, but
since we are so worried about zillions of operations per second, I hope
nobody would object if I drastically improve matching keys (several times).

Meanwhile, let's forget about it for now.

My general opinion though is that a good living code should be covered with
a good amount of tests allowing continuously refactoring it - thus keeping
it ever alive. It's not year 1985 now. We can do it. I hope.

Thanks,
-Vlad

**Comment by cliffclick** [comment]
The #1 Cause Of Bugs:

Code

Seems trite, but it's true.  I've been there, done the math - huge study 
of HotSpot's 1MLOC and the ~100K bugs in the core VM. Wrappers & 
accessors didn't help the bugs (caused several, actually).  Immutability 
didn't help (again, caused several perf issues that had to be unwound).  
Unified coding style didn't help (but generally slowed down engineers 
who's style varied from the standard).  Code reviews DO help. In depth 
comments on subtle code helped.  Shallow comments on easy code didn't help.

- Rate of bugs per-line-of-code was roughly the same between the 
best/fast/most experienced engineers vs the most junior /sloppy / 
inexperienced ones.

- Rate of bugs per-line-code was roughly 10x higher in new code than in 
code-cleanups, and again higher by 10x over bug fixes (not many fixes 
introduced a new bug).  But it clearly scaled with the total volume of 
code involved.

- Rate of code generation varied wildly, with me being a far outlying, 
and more than 2x between each engineer, down the slowest.

Since I did most of the coding overall, and also all the newest code 
(with the much higher rate) - I produced the most bugs-by-count, but was 
actually totally middle-of-the-pack on bugs-per-line-of-new-code.

Really the answer here is:
   Less Lines of Code === Less Bugs




On 10/6/2016 3:48 PM, Vlad Patryshev wrote:
> The justification here is making sure immutability makes the code 
> safer and
> does not lose a drop of performance.
>
Immutability does not make code go faster.... but replication, typically 
required by immutability, can certain make things go slower.  That's 
what's been proposed here.


>
> We can drop this specific PR, but
> since we are so worried about zillions of operations per second, I hope
> nobody would object if I drastically improve matching keys (several 
> times).
The worry here what does "drastically improve" mean, in the context of: 
a severe lack of performance understanding.
i.e. "drastically improve" ==> "drastically change" ==> things get 
slower & harder to understand - including the very important "harder to 
understand where the time goes" metric (very little to do with semantic 
correctness and a whole lot to do with performance).



>
> Meanwhile, let's forget about it for now.
>
> My general opinion though is that a good living code should be covered 
> with
> a good amount of tests allowing continuously refactoring it - thus keeping
> it ever alive. It's not year 1985 now. We can do it. I hope.

Sigh... I've been coding since long long before 1985... and I look at 
each new trend / wave / whatever roll through the coding world. I pick 
thru the good stuff, and reject what I know from 40+ yrs coding to be 
crap - even as the internet trolls loudly claim e.g. "Must use 
accessors!!! Direct field accesses are wrong!!!" and a thousand other 
drivels like that.  Most of what you hear from the Internet (and Java 
"style" trolls) is complete crap.  Won't actually help your code be more 
understandable to the next guy, or even to yourself; certainly won't 
help understand the cost model; won't help with fast multi-threading 
work (where immutability is too expensive), and so on.

I've watched code get brittle - and thus unchangable. *Tests really do 
help; I'm a firm believer. * No silver bullet though, tests cannot be 
used to prove e.g. lack of data races; currently that's only doable with 
coding discipline and culture.

I've watched slick clean fast code get bloated, ugly and slow.  The 
tests didn't help that - slow because performance wasn't tested. Witness 
the recent issue with hacking into the TCP/UDP channel vis-a-vis DL's 
training speed.  The DL training speed amounted to a poor-man's 
performance test.

H2O has long needed a really robust performance testing environment.
It doesn't have one.
Hence performance sensitive code IS fragile and canNOT be refactored - 
there's no tests of the kind that matter here.

Cliff

**Comment by vpatryshev** [comment]
Cliff,

I probably have to introduce myself. There seems to be a lot of misunderstanding here. I'll call.

Thanks,
-Vlad

**Comment by cliffclick** [comment]
Crap - I just realized I missed today.  I'm really sorry.
I'm happy to try again tomrrow if you like.

Cliff

**Comment by vpatryshev** [comment]
This is spectacular. Thank you guys for preserving this.

**Pull request closed without merging**